src/utils/PluginManager.ts

```diff
@@ -0,0 +1,123 @@
+      }
+      try {
+        if (pluginVersion != null) {
+          // this is a plugin that can be acquired from npm.
+          // it might already be installed, so check for that first.
+          const existingInstallationPath = path.join(pluginBasePath, 'node_modules', pluginName);
+          if (fs.existsSync(path.join(existingInstallationPath, 'package.json'))) {
+            // check the version number in package.json
+            const packageContents = fs.readJsonSync(
+              path.join(existingInstallationPath, 'package.json')
+            );
+            if (packageContents.version === pluginVersion) {
+              PluginManager.loadFromFilesystem([existingInstallationPath], pluginName, plugin.args);
+            } else {
+              await PluginManager.installFromNpm(pluginBasePath, plugin.name);
+              PluginManager.loadFromFilesystem([existingInstallationPath], pluginName, plugin.args);
+            }
+          } else {
+            await PluginManager.installFromNpm(pluginBasePath, plugin.name);
+            PluginManager.loadFromFilesystem([existingInstallationPath], pluginName, plugin.args);
+          }
+        } else {
+          // this is a plugin managed on the local filesystem.
+          // it may be at the base path, or in the node_modules directory
+          PluginManager.loadFromFilesystem(
+            [
+              path.join(pluginBasePath, pluginName),
+              path.join(pluginBasePath, 'node_modules', pluginName)
+            ],
+            pluginName,
+            plugin.args
+          );
+        }
+      } catch (err) {
+        logger.error(`Couldn't load plugin ${plugin.name}: ${err.message}`);
+      }
+    }
+  }
+
+  // a small test for name safety, based on typical naming patterns for npm packages
+  static isNameNpmSafe(name: string): boolean {
+    return /^[\w\d@/._-]+$/.test(name);
+  }
+
+  static async installFromNpm(installPath: string, installTarget: string) {
+    if (PluginManager.isNameNpmSafe(installTarget)) {
+      fs.ensureDirSync(installPath);
+      const installationResult = await util.promisify(execFile)(
+        'npm',
+        ['install', '--prefix', '.', installTarget],
+        {
+          cwd: installPath,
+          shell: true
+        }
+      );
+      if (installationResult.stderr.length > 0) {
+        throw new Error(installationResult.stderr);
+      } else {
+        logger.info(`Installed plugin ${installTarget}`);
+      }
+    } else {
+      throw new QuestionableNpmNameError();
+    }
+  }
+
+  static loadFromFilesystem(basePaths: string[], name: string, pluginArgs: any[] = []) {
+    for (const basePath of basePaths) {
+      const pluginPath = path.resolve(basePath);
+      if (fs.existsSync(pluginPath) && fs.statSync(pluginPath).isDirectory()) {
+        // eslint-disable-next-line @typescript-eslint/no-var-requires
+        require(pluginPath).initialize(PluginManager.registerHook, ...pluginArgs);
```

Comment:
If the plugin does not export an `initialize` function, I think I think it might be nice to provide a friendlier error here saying that the plugin does not have an initialize function.